staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/vault/vault.go

```diff
@@ -0,0 +1,200 @@
+	configFileContents, err := ioutil.ReadAll(configFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not read contents: %v", err)
+	}
+
+	var config EnvelopeConfig
+	err = yaml.Unmarshal(configFileContents, &config)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing file: %v", err)
+	}
+
+	err = checkConfig(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := newClientWrapper(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &vaultEnvelopeService{config: &config, client: client}, nil
+}
+
+func checkConfig(config *EnvelopeConfig) error {
+	if len(config.KeyNames) == 0 {
+		return fmt.Errorf("vault provider has no valid key names")
+	}
+
+	if config.Address == "" {
+		return fmt.Errorf("vault provider has no valid address")
+	}
+
+	return checkAuthConfig(config)
+}
+
+func checkAuthConfig(config *EnvelopeConfig) error {
+	var count uint
+
+	if config.Token != "" {
+		count++
+	}
+
+	if config.ClientCert != "" || config.ClientKey != "" {
+		if config.ClientCert == "" || config.ClientKey == "" {
+			return fmt.Errorf("vault provider has invalid TLS authentication information")
+		}
+		count++
+	}
+
+	if config.RoleID != "" || config.SecretID != "" {
+		if config.RoleID == "" {
+			return fmt.Errorf("vault provider has invalid approle authentication information")
+		}
+		count++
+	}
+
+	if count == 0 {
+		return fmt.Errorf("vault provider has no authentication information")
+	}
+	if count > 1 {
+		return fmt.Errorf("vault provider has more than one authentication information")
+	}
+
+	return nil
+}
+
+type vaultEnvelopeService struct {
+	config *EnvelopeConfig
+	client *clientWrapper
+}
+
+func (s *vaultEnvelopeService) Decrypt(data string) ([]byte, error) {
+	// Find the mached key
+	var key string
+	for _, name := range s.config.KeyNames {
+		if strings.HasPrefix(data, name+":") {
+			key = name
+			break
+		}
+	}
+	if key == "" {
+		return nil, fmt.Errorf("no matching vault key found")
+	}
+
+	// Replace the key name with "vault:" for Vault transit API
+	cipher := strings.Replace(data, key, "vault", 1)
+
+	//plain, _, err := s.client.decrypt(key, cipher)
+	plain, err := s.withRefreshToken((*clientWrapper).decrypt, key, cipher)
+	if err != nil {
+		return nil, err
+	}
+
+	return base64.StdEncoding.DecodeString(plain)
+}
+
+func (s *vaultEnvelopeService) Encrypt(data []byte) (string, error) {
+	// Use the frist key to encrypt
+	key := s.config.KeyNames[0]
+	plain := base64.StdEncoding.EncodeToString(data)
+
+	cipher, err := s.withRefreshToken((*clientWrapper).encrypt, key, plain)
+	if err != nil {
+		return "", err
+	}
+
+	// The format of cipher from Vault is "vault:v1:....".
+	// "vault:" is unnecessary, replace it with key name.
+	return strings.Replace(cipher, "vault", key, 1), nil
+}
+
+// The function type for clientWrapper.encrypt and clientWrapper.decrypt.
+type encryptOrDecryptFunc func(*clientWrapper, string, string) (string, error)
```

Comment:
The pattern here and `withRefreshToken` seems odd to me. Why doesn't the clientWrapper store the config that it needs to do authentication? That way you could just call encrypt/decrypt on the clientWrapper and it handles the `refreshToken` retry logic for you. It appears to me you always want to do the encrypt/decrypt action with the refresh token, so it should go down one level into the client wrapper.